### PR TITLE
ISPN-16394 Hot Rod micrometer metrics

### PR DESCRIPTION
**Pull request by pruivo**
https://issues.redhat.com/browse/ISPN-16394

### CODE REVIEW

File: commons/all/src/main/java/org/infinispan/commons/client/hotrod/metric/HotRodMetricRegistry.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+/**
+ * Hot Rod client entrypoint.
+ * <p>
+ * All metrics registered in this instance are considered global to the client. Per cache metrics are registered using
+ * {@link #withCache(String)}, where the cache's name is used to distinct the metrics. The implementation is responsible
+ * to separate each cache metric, because all cache register the same set of metrics (reads, writes, etc.).
+ *
+ * @since 15.1
+ */
+public interface HotRodMetricRegistry extends MetricRegistry {
```
Comment (ryanemerson) [inline commons/all/src/main/java/org/infinispan/commons/client/hotrod/metric/HotRodMetricRegistry.java:20]: Why are the HR specific classes in commons?

File: documentation/src/main/asciidoc/topics/code_examples/HotRodMetrics.java [hunk 2]
```diff
@@ -0,0 +1,7 @@
+HotRodMetricRegistry registry = ...;
```
Comment (ryanemerson) [inline documentation/src/main/asciidoc/topics/code_examples/HotRodMetrics.java:1]: Maybe I'm just missing it, but we should have an example in the docs of how to create the registry.

Comment (pruivo) [reply documentation/src/main/asciidoc/topics/code_examples/HotRodMetrics.java:1]: It is in the file below

File: commons/all/src/main/java/org/infinispan/commons/client/hotrod/metric/HotRodMetricRegistry.java (hunk 1 above)
Comment (pruivo) [reply commons/all/src/main/java/org/infinispan/commons/client/hotrod/metric/HotRodMetricRegistry.java:20]: To be shared between the client implementation. The alternative is to create a new module.

Comment (ryanemerson) [reply commons/all/src/main/java/org/infinispan/commons/client/hotrod/metric/HotRodMetricRegistry.java:20]: Ok, I was forgetting that we had the new implementation arriving :+1:

File: documentation/src/main/asciidoc/topics/code_examples/HotRodMetrics.java (hunk 2 above)
Comment (ryanemerson) [reply documentation/src/main/asciidoc/topics/code_examples/HotRodMetrics.java:1]: Well that's embarrassing :rofl:

Comment (ryanemerson) [reply documentation/src/main/asciidoc/topics/code_examples/HotRodMetrics.java:1]: Well that's embarrassing! :rofl:

**Review by ryanemerson — APPROVED** [review]

File: client/hotrod-client-new/src/main/java/org/infinispan/client/hotrod/impl/transport/netty/ChannelHandler.java [hunk 3]
```diff
@@ -86,6 +87,8 @@ public ChannelHandler(Configuration configuration, String sniHostName, ExecutorS
                null, ClientAuth.NONE, null, false);
          this.watcher = null;
       }
+
+      configuration.metricRegistry().createGauge("connection.pool.size", "The total number of connections", channels::size, Map.of(), null);
```
Comment (pruivo) [inline client/hotrod-client-new/src/main/java/org/infinispan/client/hotrod/impl/transport/netty/ChannelHandler.java:91]: @wburns added this gauge about the number of connections in the new client.

**Comment by pruivo** [comment]
* Added metrics to the new client.

File: client/hotrod-client-new/src/main/java/org/infinispan/client/hotrod/impl/transport/netty/ChannelHandler.java (hunk 3 above)
Comment (wburns) [reply client/hotrod-client-new/src/main/java/org/infinispan/client/hotrod/impl/transport/netty/ChannelHandler.java:91]: Sounds good

Comment (wburns) [reply client/hotrod-client-new/src/main/java/org/infinispan/client/hotrod/impl/transport/netty/ChannelHandler.java:91]: The 
      totalRetriesMetric = metricsRegistry.createCounter("connection.pool.retries", "The total number of retries", Map.of(), null);
can be added to the OperationDispatcher now. That is the class whose sole job is to dispatch operations to the correct servers so it manages retries.

Comment (pruivo) [reply client/hotrod-client-new/src/main/java/org/infinispan/client/hotrod/impl/transport/netty/ChannelHandler.java:91]: Thanks! I missed that one :+1: 
Added!

File: commons/all/src/main/java/org/infinispan/commons/stat/SimpleTimerTracker.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+   @Override
+   public void update(Duration duration) {
+      update(duration.toNanos(), TimeUnit.NANOSECONDS);
+   }
+
+   @Override
+   public void update(long value, TimeUnit timeUnit) {
+      counter.increment();
+      totalTime.add(TimeUtils.convert(value, timeUnit, TimeUnit.NANOSECONDS));
```
Comment (fax4ever) [inline commons/all/src/main/java/org/infinispan/commons/stat/SimpleTimerTracker.java:32]: is there some risk of overflow?

**Review by fax4ever — COMMENTED** [review]
Nice job! I've just some ideas... nothing critical.

File: commons/all/src/main/java/org/infinispan/commons/stat/SimpleTimerTracker.java (hunk 4 above)
Comment (pruivo) [reply commons/all/src/main/java/org/infinispan/commons/stat/SimpleTimerTracker.java:32]: eventually, the double max value is something like 1^1023 and the overflow would happen on the `sum()` method.

**Comment by pruivo** [comment]
@fax4ever updated

**Review by wburns — APPROVED** [review]
LGTM.. however CI hung for the DistSyncOnePhaseTxStateTransferTest  test which I have  never seen before and seems unrelated... I will restart another.

**Comment by fax4ever** [comment]
> @fax4ever updated

Thanks! I'm merging it, since DistSyncOnePhaseTxStateTransferTest issue seems to be not present anymore on the CI.

**Pull request merged**